backend/src/main/java/ai/verta/modeldb/cron_jobs/DeleteEntitiesCron.java

```diff
@@ -0,0 +1,666 @@
+            .append(alias)
+            .append(".")
+            .append(ModelDBConstants.DELETED)
+            .append(" = :deleted ")
+            .toString();
+    Query repositoryDeleteQuery = session.createQuery(deleteRepositorysQueryString);
+    repositoryDeleteQuery.setParameter("deleted", true);
+    LOGGER.debug("Repository delete query: {}", repositoryDeleteQuery.getQueryString());
+    List<RepositoryEntity> repositoryEntities = repositoryDeleteQuery.list();
+
+    if (!repositoryEntities.isEmpty()) {
+      for (RepositoryEntity repository : repositoryEntities) {
+        deleteRoleBindingsOfAccessibleResources(Collections.singletonList(repository));
+
+        Transaction transaction = session.beginTransaction();
+        String getRepositoryBranchesHql =
+            new StringBuilder("From ")
+                .append(BranchEntity.class.getSimpleName())
+                .append(" br where br.id.")
+                .append(ModelDBConstants.REPOSITORY_ID)
+                .append(" = :repoId ")
+                .toString();
+        Query query = session.createQuery(getRepositoryBranchesHql);
+        query.setParameter("repoId", repository.getId());
+        List<BranchEntity> branchEntities = query.list();
+
+        List<String> branches =
+            branchEntities.stream()
+                .map(branchEntity -> branchEntity.getId().getBranch())
+                .collect(Collectors.toList());
+
+        if (!branches.isEmpty()) {
+          String deleteBranchesHQL =
+              "DELETE FROM "
+                  + BranchEntity.class.getSimpleName()
+                  + " br where br.id.repository_id = :repositoryId AND br.id.branch IN (:branches)";
+          Query deleteBranchQuery = session.createQuery(deleteBranchesHQL);
+          deleteBranchQuery.setParameter("repositoryId", repository.getId());
+          deleteBranchQuery.setParameterList("branches", branches);
+          deleteBranchQuery.executeUpdate();
+        }
+
+        StringBuilder commitQueryBuilder =
+            new StringBuilder(
+                "SELECT cm FROM "
+                    + CommitEntity.class.getSimpleName()
+                    + " cm LEFT JOIN cm.repository repo WHERE repo.id = :repoId ");
+        Query<CommitEntity> commitEntityQuery =
+            session.createQuery(
+                commitQueryBuilder.append(" ORDER BY cm.date_created DESC").toString());
+        commitEntityQuery.setParameter("repoId", repository.getId());
+        List<CommitEntity> commitEntities = commitEntityQuery.list();
+
+        commitEntities.forEach(
+            commitEntity -> {
+              if (commitEntity.getRepository().contains(repository)) {
+                commitEntity.getRepository().remove(repository);
+                if (commitEntity.getRepository().isEmpty()) {
+                  session.delete(commitEntity);
+                } else {
+                  session.update(commitEntity);
+                }
+              }
+            });
```

Comment:
Commits can exist in more than one repo. The cleanup for them needs to be done more carefully. Just delete the things that are definitely associated with the repo.

Reply:
Yes, the commit will exist in multiple repositories that why I have handled that flow here. if you see line 607, I have checked that commits have a deleted repository. if yes then I am removing it from commit mapping because that commit may have second associated repository. if that commit doesn't have any repository then I will delete commit otherwise just update commit with removing deleted repository from its mapping.